### PR TITLE
allow peers to resume allocating disk buffers when their last outstan…

### PR DESCRIPTION
**Pull request by ssiloti**
…ding write completes

### CODE REVIEW

**Comment by arvidn** [comment]
I take it this solves the problem? was the previous patch also correctly referring to the m_outstanding_writing_bytes? (or did it use m_outstanding_bytes?)

**Comment by ssiloti** [comment]
Yes this solves the problem. I was mistaken in thinking that 08bac479befdc9b1dadb2989837b1bb10b650be5 referred to m_outstanding_bytes, it does correctly use m_outstanding_writing_bytes. That commit just wasn't a complete solution because it didn't add the check you see here to clear bw_disk once m_outstanding_writing_bytes drops to zero.

File: src/peer_connection.cpp [hunk 1]
```diff
@@ -2993,6 +2993,15 @@ namespace libtorrent
 
 		TORRENT_ASSERT(m_outstanding_writing_bytes >= 0);
 
+		// every peer is entitled to allocate a disk buffer if it has no writes outstanding
+		// see the comment in incoming_piece
+		if (m_outstanding_writing_bytes == 0
+			&& m_channel_state[download_channel] & peer_info::bw_disk)
+		{
+			m_counters.inc_stats_counter(counters::num_peers_down_disk, -1);
+			m_channel_state[download_channel] &= ~peer_info::bw_disk;
```
Comment (arvidn) [inline src/peer_connection.cpp]: one thing I would be a little bit concerned about with this is that there may be other assumptions associated with the bw_disk bit being set. I believe it being set means that we expect to get a callback from the disk thread once there's enough space in the disk cache. And perhaps there's an assert in that callback to make sure this bit is still set. In the case where the cache size is 0 though, we probably won't ever get that callback, and this kind of fixes up that case.

It does open up for the edge case of changing the cache size, where we may hit this _and_ get the callback at some later point. (firing the assert)

Comment (arvidn) [inline src/peer_connection.cpp]: it's not an assert, it's handled gracefully, so this should work fine

**Pull request merged**